README.md
# CycleJS examples

Derived from and inspired by CycleJS [examples](https://github.com/staltz/cycle/tree/master/examples/).
Use ES6 and JSX syntax.

Examples are grouped into lessons and placed in narrative order in tutorial-like way.
They are meant to be reviewed one by one, sequentially. Every example is accompanied
by a tutorial text (see links below). To review examples you install and run app and inspect app files.
The best way of learning is comparison. And to compare just run diff tool.

TODO setup demo instance

## Install

```
$ wget https://github.com/ivan-kleshnin/cyclejs-examples/archive/master.zip; unzip master.zip -d cyclejs-examples; rm master.zip
$ cd cyclejs-examples
$ npm install; bower install; bin/install
```

## Run

```
$ npm start -- run without watches
$ npm run devel -- run with watches
```

## Project structure
Project includes backend and frontend part. Backend is **[ExpressJS](https://github.com/strongloop/express)** + **[Nunjucks](https://github.com/mozilla/nunjucks)** based.
Frontend is obviously **[CycleJS](https://github.com/staltz/cycle)**. Backend is common for all examples. It
serves static files in simplest cases and powers API / url endpoints in more complex. Frontend consists from
independent apps with a few common files.

```
bin
  > install -- a few installation steps
  > env -- add node_modules/.bin to $PATH (required to run scripts outside of `$ npm run`)

frontend
  > 1.1-hello-cycle -- example #1.1
    > scripts -- frontend APP
    > styles -- app styles
  > ...
  > common -- common files
    > scripts -- common scripts
    > styles -- common styles

backend
  > scripts -- backend APP, API
  > templates -- nunjucks templates

shared
  > config -- config folder
```

## Examples

### Lesson 1: gentle introduction

#### Example 1.1: hello cycle
[Meet cycle](docs/lessons-1.md/#1.1)

#### Example 1.2: hello streams
[Meet streams](docs/lessons-1.md/#1.2)

#### Example 1.3: hello nodes
[Meet nodes](docs/lessons-1.md/#1.3)

## Useful links

* [Marble diagrams of Rx operators](http://rxmarbles.com/)
* [RxJS documentation](https://github.com/Reactive-Extensions/RxJS/tree/master/doc)
* [Comparison of web app architectures](https://github.com/Paqmind/reactive)


CHANGELOG.md
# Changelog

## 0.9.1

### Update
* Use CycleJS version 0.20.1

## 0.9.0

### New
* Adopt CycleJS version 0.20.0. All examples are to be rewritten.

## 0.8.0

Documentation missing

## 0.7.0

### New
* Timer example (\#1.3) with additional behavior

### Updates
* Timer examples
* Documentation

## 0.6.0

### New
* Fav examples (\#6.1, \#6.2)
* Backend

### Updates
* New project structure. Rethink folders, gulp tasks, binaries...

## 0.5.0

### New
* Menu example (\#4.3)

### Updates
* Dependencies
* Code cleanup

## 0.4.1

### Updates
* Build process

## 0.4.0

### Updates
* New project structure. Share node modules / bower components but split gulp files
  to separate builds.

## 0.3.0

### New
* Timer example (\#1.2)
* Bootstrap themes

## 0.2.0

### New
* Menu examples (\#4.1, \#4.2)
* Search example (\#5.1)
* Gulp cache layer

### Updates
* Project structure. Custom styles for every project
* Better Rx handling
* Utilize implicit returns in ES6 style
* Extend documentation

## 0.1.0
Initial version


gulpfile.js
// ENV =============================================================================================
process.env.NODE_ENV = process.env.NODE_ENV || "development";
process.env.NODE_CONFIG_DIR = process.env.NODE_CONFIG_DIR || "./shared/config";

// IMPORTS =========================================================================================
let Path = require("path");
let Glob = require("glob");
let ChildProcess = require("child_process");
let Mkdirp = require("mkdirp");
let Config = require("config");
let RunSequence = require("run-sequence");
let Gulp = require("gulp");
let JshintStylish = require("jshint-stylish");
let GulpJshint = require("gulp-jshint");
let GulpCached = require("gulp-cached");
let GulpJsx = require("gulp-jsx");
let GulpSourcemaps = require("gulp-sourcemaps");
let GulpLess = require("gulp-less");
let GulpPlumber = require("gulp-plumber");
let frontendVendors = require("./package.json").frontendVendors;

// OPTIONS =========================================================================================
let exitOnError = false;

let jsxOptions = {
  ignoreDocblock: true,
  docblockUnknownTags: true,
  unknownTagsAsString: true,
  jsx: "h",
  renameAttrs: {
    "for": "htmlFor",
    "class": "className",
  }
};

require("events").EventEmitter.defaultMaxListeners = 999;

// HELPERS =========================================================================================
function interleaveWith(array, prefix) {
  return array.reduce((memo, val) => {
    return memo.concat([prefix]).concat([val]);
  }, []);
}

// TASKS ===========================================================================================
// TODO sourcemaps
Gulp.task("build", function () {
  return Gulp.src("frontend/**/*.js")
    .pipe(GulpPlumber({errorHandler: !exitOnError}))
    .pipe(GulpCached("build"))
    .pipe(GulpJsx(jsxOptions))
    .pipe(Gulp.dest("build"));
});

Gulp.task("dist-styles", function () {
  return Gulp.src(["./frontend/**/theme.css"])
    .pipe(GulpPlumber({errorHandler: !exitOnError}))
    //.pipe(GulpLess())
    .pipe(Gulp.dest("./static/"));
});

//Gulp.task("lint", function () {
//  return Gulp.src(["./frontend/**/*.js"])
//    .pipe(GulpPlumber({errorHandler: !exitOnError}))
//    .pipe(cached("lint-react"))
//    .pipe(GulpJshint())
//    .pipe(GulpJshint.reporter(JshintStylish));
//});

Gulp.task("dist-images", function () {
  return Gulp.src(["./images/**/*"])
    .pipe(Gulp.dest("./static/images"));
});

Gulp.task("dist-vendors", function () {
  // $ browserify -d -r cyclejs [-r ...] -o ./static/scripts/vendors.js
  Mkdirp.sync("./static/common/scripts");

  let args = ["-d"]
    .concat(interleaveWith(frontendVendors, "-r"))
    .concat(["-o", "./static/common/scripts/vendors.js"]);

  let bundler = ChildProcess.spawn("browserify", args);
  bundler.stdout.pipe(process.stdout);
  bundler.stderr.pipe(process.stderr);
  bundler.on("exit", function (code) {
    if (exitOnError && code) {
      process.exit(code);
    }
  });
});

function distFactory(appCode) { // , ["predist-app"]
  return function dist() {
    let folders = Glob.sync(`build/${appCode}*`);
    if (!folders) {
      throw Error(`wrong app code ${appCode}`);
    }

    let appName = Path.basename(folders[0]);

    Mkdirp.sync(`static/${appName}/scripts`);

    // $ browserify --delay 0 -d -x cyclejs [-x ...] ./build/{app}/scripts/app.js -o ./static/{app}/scripts/app.js
    let args = ["-d", "--delay 0"]
      .concat(interleaveWith(frontendVendors, "-x"))
      .concat([`./build/${appName}/scripts/app.js`])
      .concat(["-o", `./static/${appName}/scripts/app.js`]);

    let bundler = ChildProcess.spawn("browserify", args);
    bundler.stdout.pipe(process.stdout);
    bundler.stderr.pipe(process.stderr);
    bundler.on("exit", function (code) {
      if (exitOnError && code) {
        process.exit(code);
      }
    });
  }
}

function watchifyFactory(appCode) { // , ["predist-app"]
  return function dist() {
    let folders = Glob.sync(`build/${appCode}*`);
    if (!folders) {
      throw Error(`wrong app code ${appCode}`);
    }

    let appName = Path.basename(folders[0]);

    Mkdirp.sync(`static/${appName}/scripts`);

    // $ watchify --delay 0 -v -d -x cycle [-x ...] ./build/{app}/scripts/app.js -o ./static/{app}/scripts/app.js
    let args = ["-v", "-d", "--delay 0"]
      .concat(interleaveWith(frontendVendors, "-x"))
      .concat([`./build/${appName}/scripts/app.js`])
      .concat(["-o", `./static/${appName}/scripts/app.js`]);

    let watcher = ChildProcess.spawn("watchify", args);
    watcher.stdout.pipe(process.stdout);
    watcher.stderr.pipe(process.stderr);
  }
}

// All this mess is required because of dumb Gulp v3 architecture. TODO: reconsider at Gulp v4 release (undocumented yet!)
Gulp.task("dist-1.1", distFactory("1.1"));
Gulp.task("dist-1.2", distFactory("1.2"));
Gulp.task("dist-1.3", distFactory("1.3"));
Gulp.task("dist-2.1", distFactory("2.1"));
Gulp.task("dist-2.2", distFactory("2.2"));
Gulp.task("dist-3.1", distFactory("3.1"));
Gulp.task("dist-3.2", distFactory("3.2"));
Gulp.task("dist-3.3", distFactory("3.3"));
Gulp.task("dist-3.4", distFactory("3.4"));
Gulp.task("dist-4.1", distFactory("4.1"));
Gulp.task("dist-4.2", distFactory("4.2"));
Gulp.task("dist-4.3", distFactory("4.3"));
Gulp.task("dist-5.1", distFactory("5.1"));
Gulp.task("dist-6.1", distFactory("6.1"));
Gulp.task("dist-6.2", distFactory("6.2"));

Gulp.task("watchify-1.1", watchifyFactory("1.1"));
Gulp.task("watchify-1.2", watchifyFactory("1.2"));
Gulp.task("watchify-1.3", watchifyFactory("1.3"));
Gulp.task("watchify-2.1", watchifyFactory("2.1"));
Gulp.task("watchify-2.2", watchifyFactory("2.2"));
Gulp.task("watchify-3.1", watchifyFactory("3.1"));
Gulp.task("watchify-3.2", watchifyFactory("3.2"));
Gulp.task("watchify-3.3", watchifyFactory("3.3"));
Gulp.task("watchify-3.4", watchifyFactory("3.4"));
Gulp.task("watchify-4.1", watchifyFactory("4.1"));
Gulp.task("watchify-4.2", watchifyFactory("4.2"));
Gulp.task("watchify-4.3", watchifyFactory("4.3"));
Gulp.task("watchify-5.1", watchifyFactory("5.1"));
Gulp.task("watchify-6.1", watchifyFactory("6.1"));
Gulp.task("watchify-6.2", watchifyFactory("6.2"));

Gulp.task("dist-scripts", function () {
  RunSequence(
    "dist-1.1", "dist-1.2", "dist-1.3"//,
    //"dist-2.1", "dist-2.2",
    //"dist-3.1", "dist-3.2", "dist-3.3", "dist-3.4",
    //"dist-4.1", "dist-4.2", "dist-4.3",
    //"dist-5.1",
    //"dist-6.1", "dist-6.2"
  )
});

Gulp.task("watch-build", function () {
  RunSequence(
    "watchify-1.1", "watchify-1.2", "watchify-1.3"//,
    //"watchify-2.1", "watchify-2.2",
    //"watchify-3.1", "watchify-3.2", "watchify-3.3", "watchify-3.4",
    //"watchify-4.1", "watchify-4.2", "watchify-4.3",
    //"watchify-5.1",
    //"watchify-6.1", "watchify-6.2"
  )
});

Gulp.task("watch-src", function () {
  Gulp.watch("./frontend/**/*.js", ["build"]);
  Gulp.watch("./frontend/**/*.(jpg|png|gif)", ["dist-images"]);
  Gulp.watch("./frontend/**/*.css", ["dist-styles"]);
});

Gulp.task("dist", function () {
  exitOnError = true;
  return RunSequence(
    ["build"],
    ["dist-scripts", "dist-images", "dist-styles", "dist-vendors"]
  );
});

Gulp.task("watch", ["watch-src", "watch-build"]);